core/src/main/java/edu/wpi/grip/core/MatWrapper.java

```diff
@@ -0,0 +1,391 @@
+package edu.wpi.grip.core;
+
+import org.bytedeco.javacpp.opencv_core;
+import org.bytedeco.javacpp.opencv_core.GpuMat;
+import org.bytedeco.javacpp.opencv_core.Mat;
+import org.bytedeco.javacpp.opencv_core.Size;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+import static org.bytedeco.javacpp.opencv_core.CV_16S;
+import static org.bytedeco.javacpp.opencv_core.CV_16U;
+import static org.bytedeco.javacpp.opencv_core.CV_32F;
+import static org.bytedeco.javacpp.opencv_core.CV_32S;
+import static org.bytedeco.javacpp.opencv_core.CV_64F;
+import static org.bytedeco.javacpp.opencv_core.CV_8S;
+import static org.bytedeco.javacpp.opencv_core.CV_8U;
+import static org.bytedeco.javacpp.opencv_core.IPL_DEPTH_16S;
+import static org.bytedeco.javacpp.opencv_core.IPL_DEPTH_16U;
+import static org.bytedeco.javacpp.opencv_core.IPL_DEPTH_1U;
+import static org.bytedeco.javacpp.opencv_core.IPL_DEPTH_32F;
+import static org.bytedeco.javacpp.opencv_core.IPL_DEPTH_32S;
+import static org.bytedeco.javacpp.opencv_core.IPL_DEPTH_64F;
+import static org.bytedeco.javacpp.opencv_core.IPL_DEPTH_8S;
+import static org.bytedeco.javacpp.opencv_core.IPL_DEPTH_8U;
+
+/**
+ * Wraps a GPU mat and a CPU mat and allows device memory and host memory
+ * to be used semi-transparently. A wrapper may change between wrapping an image in host memory or
+ * an image in GPU memory. A wrapper is used to minimize copies between host and device memory,
+ * which may take longer than the time savings of using a CUDA-accelerated operation.
+ *
+ * <p>Data is lazily copied between host and device memory when needed. Wrappers that
+ * are only accessed from CPU operations will never have their data stored in device memory.
+ * Accessing a wrapper from CUDA land with {@link #getGpu()} will first copy the existing data from
+ * host memory to device, if the wrapper was most recently accessed from CPU code (from either
+ * {@link #getCpu()} or {@link #rawCpu()}). This behavior also applies in the reverse, for wrappers
+ * accessed from CPU land when they have been most recently used from CUDA code.
+ */
+@SuppressWarnings("PMD.GodClass")
+public final class MatWrapper {
```

Comment:
Is this class thread safe? Does it need to be?

Reply:
No, and I don't think it needs to be. `Mat` is not thread safe and we've been fine so far with our existing synchronization constructs.